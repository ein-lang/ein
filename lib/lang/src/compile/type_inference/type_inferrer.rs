use super::{
    super::{
        error::CompileError, reference_type_resolver::ReferenceTypeResolver,
        type_canonicalizer::TypeCanonicalizer, type_equality_checker::TypeEqualityChecker,
    },
    constraint_checker::ConstraintChecker,
    constraint_collector::ConstraintCollector,
    constraint_solver::ConstraintSolver,
    variable_substitutor::VariableSubstitutor,
};
use crate::{
    ast::*,
    types::{self, Type},
};
use std::sync::Arc;

pub struct TypeInferrer {
    reference_type_resolver: Arc<ReferenceTypeResolver>,
    type_equality_checker: Arc<TypeEqualityChecker>,
    type_canonicalizer: Arc<TypeCanonicalizer>,
    constraint_collector: ConstraintCollector,
    constraint_solver: Arc<ConstraintSolver>,
}

impl TypeInferrer {
    pub fn new(
        reference_type_resolver: Arc<ReferenceTypeResolver>,
        type_equality_checker: Arc<TypeEqualityChecker>,
        type_canonicalizer: Arc<TypeCanonicalizer>,
        constraint_collector: ConstraintCollector,
        constraint_solver: Arc<ConstraintSolver>,
    ) -> Self {
        Self {
            reference_type_resolver,
            type_equality_checker,
            type_canonicalizer,
            constraint_collector,
            constraint_solver,
        }
    }

    pub fn infer(self, module: &Module) -> Result<Module, CompileError> {
        let module = module.transform_types(&mut |type_| -> Result<_, CompileError> {
            Ok(match type_ {
                Type::Unknown(unknown) => {
                    types::Variable::new(unknown.source_information().clone()).into()
                }
                _ => type_.clone(),
            })
        })?;

        let (solved_subsumption_set, mut checked_subsumption_set) =
            self.constraint_collector.collect(&module)?;

        let substitutions = self
            .constraint_solver
            .solve(solved_subsumption_set, &mut checked_subsumption_set)?;

        let substitutor = VariableSubstitutor::new(self.type_canonicalizer.clone(), substitutions);

        let checker = ConstraintChecker::new(
            substitutor.clone(),
            self.reference_type_resolver.clone(),
            self.type_equality_checker,
        );

        checker.check(checked_subsumption_set)?;

        module.transform_types(&mut |type_| substitutor.substitute(type_))
    }
}
