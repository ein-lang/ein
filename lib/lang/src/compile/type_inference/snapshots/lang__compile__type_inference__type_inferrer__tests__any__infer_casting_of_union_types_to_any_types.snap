---
source: lib/lang/src/compile/type_inference/type_inferrer.rs
expression: "infer_types(&Module::from_definitions(vec![VariableDefinition ::\n                                           new(\"x\", Boolean ::\n                                               new(true, SourceInformation ::\n                                                   dummy()), types :: Union ::\n                                               new(vec !\n                                                   [types :: Boolean ::\n                                                    new(SourceInformation ::\n                                                        dummy()) . into(),\n                                                    types :: None ::\n                                                    new(SourceInformation ::\n                                                        dummy()) . into(),],\n                                                   SourceInformation ::\n                                                   dummy(),),\n                                               SourceInformation :: dummy(),)\n                                           . into(), VariableDefinition ::\n                                           new(\"y\", Variable ::\n                                               new(\"x\", SourceInformation ::\n                                                   dummy()), types :: Any ::\n                                               new(SourceInformation ::\n                                                   dummy(),),\n                                               SourceInformation :: dummy(),)\n                                           . into()]))"
---
Ok(
    Module {
        path: ModulePath {
            package: Package {
                name: "",
                version: "",
            },
            components: [],
        },
        type_definitions: [],
        definitions: [
            VariableDefinition(
                VariableDefinition {
                    name: "x",
                    body: Boolean(
                        Boolean {
                            value: true,
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    type_: Union(
                        Union {
                            types: {
                                Boolean(
                                    Boolean {
                                        source_information: SourceInformation {
                                            source_name: "",
                                            location: Location {
                                                line_number: 0,
                                                column_number: 0,
                                            },
                                            line: "",
                                        },
                                    },
                                ),
                                None(
                                    None {
                                        source_information: SourceInformation {
                                            source_name: "",
                                            location: Location {
                                                line_number: 0,
                                                column_number: 0,
                                            },
                                            line: "",
                                        },
                                    },
                                ),
                            },
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    source_information: SourceInformation {
                        source_name: "",
                        location: Location {
                            line_number: 0,
                            column_number: 0,
                        },
                        line: "",
                    },
                },
            ),
            VariableDefinition(
                VariableDefinition {
                    name: "y",
                    body: Variable(
                        Variable {
                            name: "x",
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    type_: Any(
                        Any {
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    source_information: SourceInformation {
                        source_name: "",
                        location: Location {
                            line_number: 0,
                            column_number: 0,
                        },
                        line: "",
                    },
                },
            ),
        ],
        export: Export {
            names: {},
        },
        export_foreign: ExportForeign {
            names: {},
        },
        imports: [],
        import_foreigns: [],
    },
)
